File: algorithms/algorithms-cplusplus/leetcode/Question_0056_Merge_Intervals.cc
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
 public:
  vector<vector<int>> merge(vector<vector<int>>& intervals) {
    vector<vector<int>> res;
    if (intervals.empty()) {
      return res;
    }
    size_t n = intervals.size();
    vector<int> starts, ends;
    for (size_t i = 0; i < n; i++) {
      starts.push_back(intervals[i][0]);
      ends.push_back(intervals[i][1]);
    }
    sort(starts.begin(), starts.end());
    sort(ends.begin(), ends.end());
    for (size_t i = 0, j = 0; i < n; i++) {
      if (i == n - 1 || starts[i + 1] > ends[i]) {
        res.push_back({starts[j], ends[i]});
        j = i + 1;
      }
    }
    return res;
  }
};

File: algorithms/algorithms-python/leetcode/Question_0056_Merge_Intervals.py
#!/usr/bin/python
# coding: utf-8

class Interval(object):
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e


class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        n = len(intervals)
        starts, ends = [], []
        for i in intervals:
            starts.append(i.start)
            ends.append(i.end)
        starts.sort()
        ends.sort()

        j = 0
        res = []
        for i in range(n):
            if i == n - 1 or starts[i + 1] > ends[i]:
                res.append(Interval(starts[j], ends[i]))
                j = i + 1
        return res
